File: kotlin/src/name_processor.py
from itertools import groupby
from collections import defaultdict

def average_number_of_letters(names):
    if not names:
        return 0.0

    totalLetters = 0.0

    for name in names:
        totalLetters += len(name)

    return totalLetters/len(names)


def group_by_first_letter(names):
    if not names:
        return {}

    groups_by_name = defaultdict(list)

    for name in names:
        groups_by_name[name[0]].append(name)

    return groups_by_name


def num_names_by_first_letter(names):
    if not names:
        return {}

    names_per_letter = {}

    for name in names:
        names_per_letter[name[0]] = names_per_letter.get(name[0], 0) + 1

    return names_per_letter
  
def average_number_of_letters_immutable(names):
  if not names:
    return 0.0

  total_letters = sum([len(name) for name in names])

  return total_letters / len(names)

def group_by_first_letter_immutable(names):
  if not names:
    return {}
  
  return dict(map(lambda grouped: (grouped[0], list(grouped[1])), groupby(sorted(names), lambda name: name[0])))

def num_names_by_first_letter_immutable(names):
  if not names:
    return {}
  
  return dict(map(lambda grouped: (grouped[0], len(list(grouped[1]))), groupby(sorted(names), lambda name: name[0])))

File: ruby/src/name_processor.rb
def average_number_of_letters(names)
  return 0.0 if names.empty?
  
  total_letters = 0.0

  for name in names
    total_letters += name.length
  end

  total_letters / names.length
end

def group_by_first_letter(names)
  return {} if names.empty?

  name_groups = Hash.new([])

  for name in names
    name_groups[name[0]] += [name]
  end

  name_groups
end

def num_names_by_first_letter(names)
  return {} if names.empty?

  name_groups = Hash.new(0) 

  for name in names
    name_groups[name[0]] += 1
  end

  name_groups
end

def average_number_of_letters_immutable(names)
  return 0.0 if names.empty?

  total_letters = names.map { |name| name.length }.sum.to_f

  total_letters / names.length 
end

def group_by_first_letter_immutable(names)
  return {} if names.empty?

  names.group_by { |name| name[0] }
end

def num_names_by_first_letter_immutable(names)
  return {} if names.empty?

  names.group_by { |name| name[0] }.transform_values(&:length)
end
